simple-ipc.c

```diff
@@ -182,4 +182,143 @@ int ipc_send_command(const char *path, const char *message, struct strbuf *answe
 		close(fd);
 	return ret < 0 ? -1 : 0;
 }
+#elif !defined(NO_UNIX_SOCKETS)
+#include "unix-socket.h"
+#include "sigchain.h"
+
+static const char *fsmonitor_listener_path;
+
+static void set_socket_blocking_flag(int fd, int make_nonblocking)
+{
+	int flags;
+
+	flags = fcntl(fd, F_GETFL, NULL);
+
+	if (flags < 0)
+		die(_("fcntl failed"));
+
+	if (make_nonblocking)
+		flags |= O_NONBLOCK;
+	else
+		flags &= ~O_NONBLOCK;
+
+	if (fcntl(fd, F_SETFL, flags) < 0)
+		die(_("fcntl failed"));
+}
+
+static int reply(void *reply_data, const char *response, size_t len)
+{
+	int fd = *(int *)reply_data;
+
+	return write_packetized_from_buf(response, len, fd, 0);
+}
+
+/* in ms */
+#define LISTEN_TIMEOUT 50000
+#define RESPONSE_TIMEOUT 1000
+
+static void unlink_listener_path(void)
+{
+	if (fsmonitor_listener_path)
+		unlink(fsmonitor_listener_path);
+}
+
+int ipc_listen_for_commands(struct ipc_command_listener *listener)
+{
+	int ret = 0, fd;
+
+	fd = unix_stream_listen(listener->path);
+	if (fd < 0)
+		return error_errno(_("could not set up socket for %s"),
+				   listener->path);
+
+	fsmonitor_listener_path = listener->path;
+	atexit(unlink_listener_path);
+
+	trace2_region_enter("simple-ipc", "listen", the_repository);
```

Comment:
this trace2 region is missing from the Windows implementation. Can we add it there? It would be nice to see what we can do to make these as equivalent as possible?

Perhaps we need to break this method into sub-parts that have platform-specific implementations, but the general structure is platform agnostic?

Reply:
I like the idea of making this as platform-agnostic as possible, but when I look through the code, it does not seem to lend itself to being abstracted, unless we want to pay a penalty (which I am not really prepared).

About that Trace2 region: absolutely 👍 . I also fail to miss the fact that the `_leave()` call is missing.